core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1266,6 +1266,9 @@ class DAGScheduler(
             }
             if (failedEpoch.contains(execId) && smt.epoch <= failedEpoch(execId)) {
               logInfo(s"Ignoring possibly bogus $smt completion from executor $execId")
+            } else if (failedStages.contains(shuffleStage)) {
```

Comment:
Why we only have a problem with shuffle map task not result task?

Reply:
This also confuse me before, as far as I'm concerned, the result task in such scenario(speculative task fail but original task success) is ok because it has no child stage, we can use the success task's result and `markStageAsFinished`. But for shuffle map task, it will cause inconformity between mapOutputTracker and stage's pendingPartitions, it must fix.
I'm not sure of ResultTask's behavior, can you give some advice?

Reply:
ResultStage usually has children. e.g. rdd1 depends on rdd2 with a shuffle dependency. Then the job has 2 stages, the result stage depends on the shuffle map stage.

Reply:
Sorry I may nitpick kere. Can you simulate what happens to result task if FechFaileded comes before task success?

Reply:
seems we may mistakenly mark a job as finished?

Reply:
>Sorry I may nitpick here.

No, that's necessary, I should have to make sure about this, thanks for your advice! :)

> Can you simulate what happens to result task if FechFaileded comes before task success?

Sure, but it maybe hardly to reproduce this in real env, I'll try to fake it on UT first ASAP.

Reply:
Added the UT for simulating this scenario happens to result task.